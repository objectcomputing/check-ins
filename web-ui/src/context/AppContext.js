import React, { useEffect, useReducer, useMemo } from "react";
import { createSelector } from 'reselect';
import { getCurrentUser, updateMember, getAllMembers } from "../api/member";
import { getCheckinByMemberId, createCheckin } from "../api/checkins";
import { BASE_API_URL } from "../api/api";
import axios from "axios";

export const MY_PROFILE_UPDATE = "@@check-ins/update_profile";
export const UPDATE_USER_BIO = "@@check-ins/update_bio";
export const UPDATE_CHECKINS = "@@check-ins/update_checkins";
export const UPDATE_INDEX = "@@check-ins/update_index";
export const UPDATE_TOAST = "@@check-ins/update_toast";
export const UPDATE_CURRENT_CHECKIN = "@@check-ins/update_current_checkin";
export const UPDATE_TEAMS = "@@check-ins/update_teams";
export const UPDATE_MEMBER_PROFILES = "@@check-ins/update_member_profiles";
export const UPDATE_TEAM_MEMBERS = "@@check-ins/update_team_members";
export const UPDATE_SELECTED_PROFILE = "@@check-ins/update_selected_profile";
const SET_CSRF = "@@check-ins/update_csrf";

const AppContext = React.createContext();

const reducer = (state, action) => {
  switch (action.type) {
    case MY_PROFILE_UPDATE:
      state.userProfile = action.payload;
      break;
    case UPDATE_USER_BIO:
      state.userProfile.memberProfile.bioText = action.payload;
      updateMember(state.userProfile.memberProfile);
      break;
    case UPDATE_CHECKINS:
      state.checkins = action.payload;
      //sort by date
      state.checkins.sort(function (a, b) {
        return new Date(...a.checkInDate) - new Date(...b.checkInDate);
      });
      state.currentCheckin = state.checkins[state.checkins.length - 1];
      break;
    case SET_CSRF:
      state.csrf = action.payload;
      break;
    case UPDATE_TOAST:
      state.toast = action.payload;
      break;
    case UPDATE_TEAMS:
      state.teams = action.payload;
      break;
    case UPDATE_MEMBER_PROFILES:
      state.memberProfiles = action.payload;
      break;
    case UPDATE_TEAM_MEMBERS:
      state.teamMembers = action.payload;
      break;
    case UPDATE_CURRENT_CHECKIN:
      state.currentCheckin = action.payload;
      break;
    case UPDATE_SELECTED_PROFILE:
      const { payload } = action;
      if (state.selectedProfile !== payload) {
        state.selectedProfile = payload;
        state.currentCheckin = payload ? payload.checkIn : {};
      }
      if (payload === undefined) {
        state.checkins = [];
      }
      break;
    default:
  }
  return { ...state };
};

const initialState = {
  checkins: [],
  csrf: undefined,
  currentCheckin: {},
  teams: [],
  memberProfiles: [],
  index: 0,
  selectedProfile: undefined,
  toast: {
    severity: "",
    toast: "",
  },
  userProfile: undefined,
};

const getCheckins = async (id, pdlId, date, dispatch, csrf) => {
  const res = await getCheckinByMemberId(id, csrf);
  let data =
    res.payload && res.payload.data && res.payload.status === 200 && !res.error
      ? res.payload.data
      : null;
  if (data && data.length > 0) {
    const allComplete = data.every((checkin) => checkin.completed === true);
    if (allComplete) {
      const prevCheckinDate = data[data.length - 1].checkInDate;
      if (pdlId) {
        const res = await createCheckin({
          teamMemberId: id,
          pdlId: pdlId,
          checkInDate: date(3, prevCheckinDate),
          completed: false,
        });
        const checkin =
          res.payload && res.payload.data && !res.error
            ? res.payload.data
            : null;
        data.push(checkin);
      }
    }
  } else if (data && data.length === 0) {
    if (pdlId) {
      const res = await createCheckin({
        teamMemberId: id,
        pdlId: pdlId,
        checkInDate: date(1),
        completed: false,
      });
      const checkin =
        res.payload && res.payload.data && !res.error ? res.payload.data : null;
      data = [checkin];
    }
  }
  //without this check you get infinite checkin calls
  if (data.length > 0) {
    dispatch({ type: UPDATE_CHECKINS, payload: data });
  }
};

const AppContextProvider = (props) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const memberProfile =
    state && state.userProfile && state.userProfile.memberProfile
      ? state.userProfile.memberProfile
      : undefined;
  const id = memberProfile ? memberProfile.id : undefined;
  const selectedProfile = state && state.selectedProfile;
  const selectedId = selectedProfile ? selectedProfile.id : undefined;


  const { csrf } = state;

  useEffect(() => {
    const getCsrf = async () => {
      if (!csrf) {
        const res = await axios({
          url: `${BASE_API_URL}/csrf/cookie`,
          responseType: "text",
          withCredentials: true,
        });
        if (res && res.data) {
          dispatch({ type: SET_CSRF, payload: res.data });
        }
      }
    };
    getCsrf();
  }, [csrf]);

  useEffect(() => {
    const updateUserProfile = async () => {
      let res = await getCurrentUser(csrf);
      let profile =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profile) {
        dispatch({ type: MY_PROFILE_UPDATE, payload: profile });
      }
    };
    if (csrf) {
      updateUserProfile();
    }
  }, [csrf]);

  const date = (months, prevCheckinDate) => {
    let currentMonth = new Date().getMonth() + 1;
    let newDate = prevCheckinDate ? new Date(...prevCheckinDate) : new Date();
    if (prevCheckinDate) {
      let prevCheckinMonth = prevCheckinDate.getMonth() + 1;
      newDate.setMonth(
        newDate.getMonth() + currentMonth - prevCheckinMonth >= 3 ? 1 : months
      );
    } else {
      newDate.setMonth(newDate.getMonth() + months);
    }
    const year = newDate.getFullYear();
    const month = newDate.getMonth() + 1;
    const day = newDate.getDate();
    const hours = newDate.getHours();
    const minutes = newDate.getMinutes();
    const dateTimeArray = [year, month, day, hours, minutes, 0];
    return dateTimeArray;
  };

  useEffect(() => {
    async function getMemberProfiles() {
      let res = await getAllMembers(csrf);
      let profiles =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profiles) {
        dispatch({ type: UPDATE_MEMBER_PROFILES, payload: profiles });
      }
    }
    if (csrf) {
      getMemberProfiles();
    }
  }, [csrf]);

  useEffect(() => {
    if (selectedId && csrf) {
      getCheckins(selectedId, id, date, dispatch, csrf);
    }
  }, [csrf, selectedId, id]);

  const value = useMemo(() => {
    return { state, dispatch };
  }, [state]);
  return (
    <AppContext.Provider value={props.value || value}>
      {props.children}
    </AppContext.Provider>
  );
};

export const selectMemberProfiles = state => state.memberProfiles;
export const selectTeamMembers = state => state.teamMembers;

export const selectProfileMap = createSelector(
  selectMemberProfiles,
  (memberProfiles) => {
    if (memberProfiles && memberProfiles.length) {
      memberProfiles = memberProfiles.reduce((mappedById, profile) => {
        mappedById[profile.id] = profile;
        return mappedById;
      }, {});
    }
    return memberProfiles;
  }
);

export const selectMembersByTeamId = id => createSelector(
  selectTeamMembers,
  (teamMembers) => {
    let members = [];
    if (teamMembers && teamMembers.length) {
      members = teamMembers.filter((member) => member.teamid === id);
    }
    return members;
  }
);

export const selectMemberProfilesByTeamId = id => createSelector(
  selectMembersByTeamId(id),
  selectProfileMap,
  (members, profileMap) => members.map((member) => {
      return { ...profileMap[member.memberid], ...member };
    })
);

export { AppContext, AppContextProvider };
