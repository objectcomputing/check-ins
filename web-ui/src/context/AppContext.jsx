import React, { useEffect, useReducer, useMemo } from 'react';
import { reducer, initialState } from './reducer';
import { getCheckins, getAllCheckinsForAdmin } from './thunks';
import {
  MY_PROFILE_UPDATE,
  SET_CSRF,
  SET_ROLES,
  SET_USER_ROLES,
  UPDATE_GUILDS,
  UPDATE_MEMBER_SKILLS,
  UPDATE_MEMBER_PROFILES,
  UPDATE_TERMINATED_MEMBERS,
  UPDATE_SKILLS,
  UPDATE_CERTIFICATIONS,
  UPDATE_TEAMS,
  UPDATE_PEOPLE_LOADING,
  UPDATE_TEAMS_LOADING
} from './actions';
import {
  getCurrentUser,
  getAllMembers,
  getAllTerminatedMembers
} from '../api/member';
import {
  selectCanViewCheckinsPermission,
  selectCanViewTerminatedMembers,
} from './selectors';
import { getAllRoles, getAllUserRoles } from '../api/roles';
import { getMemberSkills } from '../api/memberskill';
import { BASE_API_URL } from '../api/api';
import { getAllGuilds } from '../api/guild';
import { getSkills } from '../api/skill';
import { getAllTeams } from '../api/team';
import {getCertifications} from "../api/certification.js";

const AppContext = React.createContext();

function getSessionCookieValue(name) {
  const cookies = document?.cookie?.split(';');
  for (let i = 0; i < cookies.length; i++) {
    const cookie = cookies[i].trim();
    if (cookie.startsWith(name + '=')) {
      return decodeURIComponent(cookie.substring(name.length + 1));
    }
  }
  return null;
}

const AppContextProvider = props => {
  const [state, dispatch] = useReducer(
    reducer,
    props?.value?.state || initialState
  );
  const userProfile =
    state && state.userProfile ? state.userProfile : undefined;
  const memberProfile =
    userProfile && userProfile.memberProfile
      ? userProfile.memberProfile
      : undefined;

  const id = memberProfile ? memberProfile.id : undefined;
  const pdlId = memberProfile ? memberProfile.pdlId : undefined;
  const {
    csrf,
    guilds,
    teams,
    memberSkills,
    memberProfiles,
    checkins,
    skills,
    certifications,
    roles,
    userRoles
  } = state;
  const url = `${BASE_API_URL}/csrf/cookie`;
  useEffect(() => {
    const getCsrf = async () => {
      if (!csrf) {
        const payload = getSessionCookieValue('_csrf');
        if (payload) {
          dispatch({ type: SET_CSRF, payload });
        } else {
          const res = await fetch(url, {
            responseType: 'text',
            credentials: 'include'
          });
          if (res && res.ok) {
            dispatch({ type: SET_CSRF, payload: await res.text() });
          }
        }
      }
    };
    getCsrf();
  }, [csrf]);

  useEffect(() => {
    async function getGuilds() {
      let res = await getAllGuilds(csrf);
      let data =
        res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
          ? res.payload.data
          : null;
      if (data) {
        dispatch({ type: UPDATE_GUILDS, payload: data });
      }
    }
    if (csrf && !guilds) {
      getGuilds();
    }
  }, [csrf, guilds]);

  useEffect(() => {
    async function getTeams() {
      let res = await getAllTeams(csrf);
      let data =
        res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
          ? res.payload.data
          : null;
      if (data) {
        dispatch({ type: UPDATE_TEAMS, payload: data });
        dispatch({ type: UPDATE_TEAMS_LOADING });
      }
    }
    if (csrf && !teams) {
      dispatch({ type: UPDATE_TEAMS_LOADING });
      getTeams();
    }
  }, [csrf, teams, dispatch]);

  useEffect(() => {
    const updateUserProfile = async () => {
      let res = await getCurrentUser(csrf);
      let profile =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profile) {
        dispatch({ type: MY_PROFILE_UPDATE, payload: profile });
      }
    };
    if (csrf && !userProfile) {
      updateUserProfile();
    }
  }, [csrf, userProfile]);

  useEffect(() => {
    const getAllMemberSkills = async () => {
      const res = await getMemberSkills(csrf);
      const memberSkills =
        res && res.payload && res.payload.data ? res.payload.data : null;
      if (memberSkills) {
        dispatch({ type: UPDATE_MEMBER_SKILLS, payload: memberSkills });
      }
    };
    if (csrf && !memberSkills) {
      getAllMemberSkills();
    }
  }, [csrf, memberSkills]);

  useEffect(() => {
    async function getMemberProfiles() {
      let res = await getAllMembers(csrf);
      let profiles =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profiles) {
        dispatch({ type: UPDATE_MEMBER_PROFILES, payload: profiles });
        dispatch({ type: UPDATE_PEOPLE_LOADING, payload: false });
      }
    }
    async function getTerminatedMembers() {
      let res = await getAllTerminatedMembers(csrf);
      let profiles =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profiles) {
        dispatch({ type: UPDATE_TERMINATED_MEMBERS, payload: profiles });
      }
    }
    if (csrf && userProfile && !memberProfiles) {
      dispatch({ type: UPDATE_PEOPLE_LOADING, payload: true });
      getMemberProfiles();
      if (selectCanViewTerminatedMembers(state)) {
        getTerminatedMembers();
      }
    }
  }, [csrf, userProfile, memberProfiles]);

  useEffect(() => {
    function getAllTheCheckins() {
      if (
        userProfile &&
        userProfile.permissions?.some(p =>
          p?.permission?.includes('CAN_VIEW_CHECKINS_REPORT')
        ) &&
        id &&
        csrf
      ) {
        getAllCheckinsForAdmin(dispatch, csrf);
      } else if (id && csrf && selectCanViewCheckinsPermission(state)) {
        getCheckins(id, pdlId, dispatch, csrf);
      }
    }
    if (csrf && !checkins) {
      getAllTheCheckins();
    }
  }, [csrf, pdlId, id, userProfile, checkins]);

  useEffect(() => {
    const getAllSkills = async () => {
      const res = await getSkills(csrf);
      const data =
        res &&
        res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
          ? res.payload.data
          : null;
      if (data && data.length > 0) {
        dispatch({ type: UPDATE_SKILLS, payload: data });
      }
    };
    if (csrf && !skills) {
      getAllSkills();
    }
  }, [csrf, skills]);

  useEffect(() => {
    const getAllCertifications = async () => {
      const res = await getCertifications(csrf);
      const data =
          res &&
          res.payload &&
          res.payload.data &&
          res.payload.status === 200 &&
          !res.error
              ? res.payload.data
              : null;
      if (data && data.length > 0) {
        dispatch({ type: UPDATE_CERTIFICATIONS, payload: data });
      }
    };
    if (csrf && !certifications) {
      getAllCertifications();
    }
  }, [csrf, certifications]);

  useEffect(() => {
    const getRoles = async () => {
      const res = await getAllRoles(csrf);
      const data =
        res &&
        res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
          ? res.payload.data
          : null;
      if (data && Array.isArray(data) && data.length > 0) {
        dispatch({ type: SET_ROLES, payload: data });
      }
    };
    if (csrf && !roles) {
      getRoles();
    }
  }, [csrf, roles]);

  useEffect(() => {
    const getUserRoles = async () => {
      // make call to the API
      let res = await getAllUserRoles(csrf);
      return res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
        ? res.payload.data
        : null;
    };

    if (csrf && !userRoles) {
      getUserRoles().then(userRoles => {
        dispatch({ type: SET_USER_ROLES, payload: userRoles });
      });
    }
  }, [csrf, userRoles]);

  const value = useMemo(() => {
    return { state, dispatch };
  }, [state]);
  return (
    <AppContext.Provider value={props.value || value}>
      {props.children}
    </AppContext.Provider>
  );
};

export { AppContext, AppContextProvider };
