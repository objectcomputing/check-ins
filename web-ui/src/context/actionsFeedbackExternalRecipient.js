export const SET_CSRF = '@@check-ins/update_csrf';
export const UPDATE_TOAST = '@@check-ins/update_toast';

