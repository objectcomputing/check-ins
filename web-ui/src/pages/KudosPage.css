.kudos-page {
  margin: 3rem 4rem 3rem 4rem;
}

.kudos-page .kudos-title {
  display: flex;
  justify-content: center;
}

.kudos-page .received-kudos-header,
.kudos-page .sent-kudos-header {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin-top: 2rem;
}

.kudos-page .empty-kudos-container {
  width: 100%;
  text-align: center;
  background-color: #e5e5e5;
  color: gray;
  border-radius: 4px;
  padding: 1rem 0;
  margin-top: 1rem;
}

.kudos-page .kudos-card {
  margin-top: 1rem;
}

.kudos-page .kudos-page-header {
  display: flex;
  justify-content: space-between;
}
