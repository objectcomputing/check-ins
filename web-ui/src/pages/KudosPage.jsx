import React, { useCallback, useContext, useEffect, useState } from "react";
import { useParams } from 'react-router-dom';
import { styled } from "@mui/material/styles";
import { Button, Tab, Typography } from "@mui/material";
import MenuItem from '@mui/material/MenuItem';
import FormControl from '@mui/material/FormControl';
import TextField from '@mui/material/TextField';
import { TabContext, TabList, TabPanel } from "@mui/lab";
import { AppContext } from "../context/AppContext";
import {
  selectCsrfToken,
  selectCurrentUser,
  selectHasCreateKudosPermission,
} from "../context/selectors";
import { getReceivedKudos, getSentKudos, getAllKudos } from "../api/kudos";
import { UPDATE_TOAST } from "../context/actions";
import KudosCard from "../components/kudos_card/KudosCard";

import "./KudosPage.css";
import KudosDialog from "../components/kudos_dialog/KudosDialog";
import StarIcon from "@mui/icons-material/Star";

import ArchiveIcon from "@mui/icons-material/Archive";
import UnarchiveIcon from "@mui/icons-material/Unarchive";

import SkeletonLoader from "../components/skeleton_loader/SkeletonLoader";

const PREFIX = "KudosPage";
const classes = {
  expandOpen: `${PREFIX}-expandOpen`,
  expandClose: `${PREFIX}-expandClose`,
};

const Root = styled("div")({
  [`& .${classes.expandOpen}`]: {
    transform: "rotate(180deg)",
    transition: "transform 0.1s linear",
    marginLeft: "auto",
  },
  [`& .${classes.expandClose}`]: {
    transform: "rotate(0deg)",
    transition: "transform 0.1s linear",
    marginLeft: "auto",
  },
});

const validTabName = (name) => {
  switch (name) {
    case "received":
    case "sent":
    case "public":
      break;
    default:
      name && console.warn(`Invalid tab: ${name}`);
      name = "received";
  }
  return name;
}

const DateRange = {
  THREE_MONTHS: '3mo',
  SIX_MONTHS: '6mo',
  ONE_YEAR: '1yr',
  ALL_TIME: 'all'
};

const KudosPage = () => {
  const { initialTab } = useParams();
  const { state, dispatch } = useContext(AppContext);
  const csrf = selectCsrfToken(state);

  const currentUser = selectCurrentUser(state);

  const [kudosDialogOpen, setKudosDialogOpen] = useState(false);
  const [kudosTab, setKudosTab] = useState(validTabName(initialTab));
  const [receivedKudos, setReceivedKudos] = useState([]);
  const [receivedKudosLoading, setReceivedKudosLoading] = useState(true);
  const [sentKudos, setSentKudos] = useState([]);
  const [sentKudosLoading, setSentKudosLoading] = useState(true);
  const [publicKudos, setPublicKudos] = useState([]);
  const [publicKudosLoading, setPublicKudosLoading] = useState(true);
  const [dateRange, setDateRange] = useState(DateRange.THREE_MONTHS);

  const isInRange = (requestDate) => {
    const oldestDate = new Date();
    switch (dateRange) {
      case DateRange.SIX_MONTHS:
        oldestDate.setMonth(oldestDate.getMonth() - 6);
        break;
      case DateRange.ONE_YEAR:
        oldestDate.setFullYear(oldestDate.getFullYear() - 1);
        break;
      case DateRange.ALL_TIME:
        return true;
      case DateRange.THREE_MONTHS:
      default:
        oldestDate.setMonth(oldestDate.getMonth() - 3);
    }

    if (Array.isArray(requestDate)) {
      requestDate = new Date(requestDate.join('/'));
      // have to do for Safari
    }
    return requestDate >= oldestDate;
  };

  const loadReceivedKudos = useCallback(async () => {
    setReceivedKudosLoading(true);
    const res = await getReceivedKudos(currentUser.id, csrf);
    if (res?.payload?.data && !res.error) {
      setReceivedKudosLoading(false);
      return res.payload.data.filter((k) => isInRange(k.dateCreated));
    }
  }, [csrf, dispatch, currentUser.id, dateRange]);

  const loadSentKudos = useCallback(async () => {
    setSentKudosLoading(true);
    const res = await getSentKudos(currentUser.id, csrf);
    if (res?.payload?.data && !res.error) {
      setSentKudosLoading(false);
      return res.payload.data.filter((k) => isInRange(k.dateCreated));
    }
  }, [csrf, dispatch, currentUser.id, dateRange]);

  const loadPublicKudos = useCallback(async () => {
    setPublicKudosLoading(true);
    const res = await getAllKudos(csrf);
    if (res?.payload?.data && !res.error) {
      setPublicKudosLoading(false);
      return res.payload.data.filter((k) => isInRange(k.dateCreated));
    }
  }, [csrf, dispatch, currentUser.id, dateRange]);

  const loadAndSetReceivedKudos = () => {
    loadReceivedKudos().then((data) => {
      if (data) {
        const filtered = data.filter((kudo) =>
          kudo.recipientMembers.some((member) => member.id === currentUser.id)
        );
        setReceivedKudos(filtered);
      }
    });
  };

  const loadAndSetSentKudos = () => {
    loadSentKudos().then((data) => {
      if (data) {
        setSentKudos(data.filter((kudo) => kudo.senderId === currentUser.id));
      }
    });
  };

  const loadAndSetPublicKudos = () => {
    loadPublicKudos().then((data) => {
      if (data) {
        setPublicKudos(data);
      }
    });
  };

  useEffect(() => {
    if (csrf && currentUser && currentUser.id) {
      loadAndSetReceivedKudos();
      loadAndSetSentKudos();
      loadAndSetPublicKudos();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [csrf, currentUser, kudosTab, dateRange]);

  const handleTabChange = useCallback(
    (event, newTab) => {
      setKudosTab(validTabName(newTab));
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [loadReceivedKudos, loadSentKudos]
  );

  return (
    <Root className="kudos-page">
      <div className="kudos-page-header">
        <KudosDialog
          open={kudosDialogOpen}
          onClose={() => setKudosDialogOpen(false)}
        />
        <FormControl className={classes.textField}>
          <TextField
            id="select-time"
            select
            fullWidth
            size="small"
            label="Show kudos within"
            onChange={e => setDateRange(e.target.value)}
            value={dateRange}
            variant="outlined"
          >
            <MenuItem value={DateRange.THREE_MONTHS}>Past 3 months</MenuItem>
            <MenuItem value={DateRange.SIX_MONTHS}>Past 6 months</MenuItem>
            <MenuItem value={DateRange.ONE_YEAR}>Past year</MenuItem>
            <MenuItem value={DateRange.ALL_TIME}>All time</MenuItem>
          </TextField>
        </FormControl>
        {selectHasCreateKudosPermission(state) && <Button
          className="kudos-dialog-open"
          variant="outlined"
          startIcon={<StarIcon />}
          onClick={() => setKudosDialogOpen(true)}
        >
          Give Kudos
        </Button>}
      </div>
      <TabContext value={kudosTab}>
        <div className="kudos-tab-container">
          <TabList onChange={handleTabChange}>
            <Tab
              label="Received"
              value="received"
              icon={<ArchiveIcon />}
              iconPosition="start"
            />
            <Tab
              label="Sent"
              value="sent"
              icon={<UnarchiveIcon />}
              iconPosition="start"
            />
            <Tab
              label="Public"
              value="public"
              icon={<StarIcon />}
              iconPosition="start"
            />
          </TabList>
        </div>
        <TabPanel value="received" style={{ padding: "1rem 0" }}>
          {receivedKudosLoading ? (
            Array.from({ length: 5 }).map((_, index) => (
              <SkeletonLoader key={index} type="kudos" />
            ))
          ) : receivedKudos.length > 0 &&
            receivedKudos.filter(
              (kudo, index) =>
                kudo.recipientMembers[index]?.id === currentUser.id
            ) ? (
            <div className="received-kudos-list">
              {receivedKudos.map((k) => (
                <KudosCard
                  key={k.id}
                  kudos={k}
                  onKudosAction={loadAndSetReceivedKudos}
                />
              ))}
            </div>
          ) : (
            <div className="no-pending-kudos-message">
              <Typography variant="body2">
                There are currently no received kudos
              </Typography>
            </div>
          )}
        </TabPanel>
        <TabPanel value="sent" style={{ padding: "1rem 0" }}>
          {sentKudosLoading ? (
            Array.from({ length: 5 }).map((_, index) => (
              <SkeletonLoader key={index} type="kudos" />
            ))
          ) : sentKudos.length > 0 ? (
            <div>
              {sentKudos.map((k) => (
                <KudosCard key={k.id} kudos={k} includeEdit
                  onKudosAction={loadAndSetSentKudos}/>
              ))}
            </div>
          ) : (
            <div className="no-sent-kudos-message">
              <Typography variant="body2">
                There are currently no sent kudos
              </Typography>
            </div>
          )}
        </TabPanel>
        <TabPanel value="public" style={{ padding: "1rem 0" }}>
          {publicKudosLoading ? (
            Array.from({ length: 5 }).map((_, index) => (
              <SkeletonLoader key={index} type="kudos" />
            ))
          ) : publicKudos.length > 0
            ? (
            <div className="received-kudos-list">
              {publicKudos.map((k) => (
                <KudosCard
                  key={k.id}
                  kudos={k}
                />
              ))}
            </div>
          ) : (
            <div className="no-pending-kudos-message">
              <Typography variant="body2">
                There are currently no public kudos
              </Typography>
            </div>
          )}
        </TabPanel>
      </TabContext>
    </Root>
  );
};

export default KudosPage;
