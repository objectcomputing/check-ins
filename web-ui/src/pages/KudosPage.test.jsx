import React from 'react';
import KudosPage from './KudosPage';
import { AppContextProvider } from '../context/AppContext';
import { http, HttpResponse } from 'msw';
import { setupServer } from 'msw/node';
import { BrowserRouter } from 'react-router-dom';

window.snackDispatch = vi.fn();

const userStateWithPermission = {
  state: {
    csrf: 'csrf',
    userProfile: {
      id: 'current-id',
      name: 'Mitch Hedberg',
      role: ['MEMBER'],
      permissions: [{ permission: '' }],
    },
    teams: [],
    memberProfiles: [
      {
        id: 'some-id',
        firstName: 'First',
        lastName: 'Last',
      },
      {
        id: 'other-id',
        firstName: 'First',
        lastName: 'Last',
      },
    ],
  }
};

const server = setupServer(
  http.get('http://localhost:8080/services/kudos', ({ request }) => {
    return HttpResponse.json([
      {
        id: 'kudos-id',
        message: 'You are great!',
        senderId: 'some-id',
        recipientTeam: null,
        dateCreated: [2024, 11, 11],
        dateApproved: [2024, 11, 11],
        recipientMembers: [
          {id: 'current-id'},
        ],
      },
    ]);
  }),
);

beforeAll(() => server.listen({ onUnhandledRequest(request, print) {} }));
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

it('renders correctly', () => {
  snapshot(
    <AppContextProvider value={userStateWithPermission}>
      <BrowserRouter>
        <KudosPage />
      </BrowserRouter>
    </AppContextProvider>
  );
});
