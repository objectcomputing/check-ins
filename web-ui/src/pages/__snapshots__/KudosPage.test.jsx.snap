// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly 1`] = `
<div>
  <div
    class="kudos-page css-177onil"
  >
    <div
      class="kudos-page-header"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary kudos-dialog-open css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="StarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
            />
          </svg>
        </span>
        Give Kudos
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="kudos-tab-container"
    >
      <div
        class="MuiTabs-root css-1ujnqem-MuiTabs-root"
      >
        <div
          class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
          style="overflow: hidden; margin-bottom: 0px;"
        >
          <div
            class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
            role="tablist"
          >
            <button
              aria-controls="mui-p-12978-P-RECEIVED"
              aria-selected="true"
              class="MuiButtonBase-root MuiTab-root MuiTab-labelIcon MuiTab-textColorPrimary Mui-selected css-h0q0iv-MuiButtonBase-root-MuiTab-root"
              id="mui-p-12978-T-RECEIVED"
              role="tab"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTab-iconWrapper css-i4bv87-MuiSvgIcon-root"
                data-testid="ArchiveIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m20.54 5.23-1.39-1.68C18.88 3.21 18.47 3 18 3H6c-.47 0-.88.21-1.16.55L3.46 5.23C3.17 5.57 3 6.02 3 6.5V19c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V6.5c0-.48-.17-.93-.46-1.27M12 17.5 6.5 12H10v-2h4v2h3.5zM5.12 5l.81-1h12l.94 1z"
                />
              </svg>
              Received
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-controls="mui-p-12978-P-SENT"
              aria-selected="false"
              class="MuiButtonBase-root MuiTab-root MuiTab-labelIcon MuiTab-textColorPrimary css-h0q0iv-MuiButtonBase-root-MuiTab-root"
              id="mui-p-12978-T-SENT"
              role="tab"
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTab-iconWrapper css-i4bv87-MuiSvgIcon-root"
                data-testid="UnarchiveIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m20.55 5.22-1.39-1.68C18.88 3.21 18.47 3 18 3H6c-.47 0-.88.21-1.15.55L3.46 5.22C3.17 5.57 3 6.01 3 6.5V19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6.5c0-.49-.17-.93-.45-1.28M12 9.5l5.5 5.5H14v2h-4v-2H6.5zM5.12 5l.82-1h12l.93 1z"
                />
              </svg>
              Sent
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <span
            class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
            style="left: 0px; width: 0px;"
          />
        </div>
      </div>
    </div>
    <div
      aria-labelledby="mui-p-12978-T-RECEIVED"
      class="MuiTabPanel-root css-13xfq8m-MuiTabPanel-root"
      id="mui-p-12978-P-RECEIVED"
      role="tabpanel"
      style="padding: 1rem 0px;"
    />
    <div
      aria-labelledby="mui-p-12978-T-SENT"
      class="MuiTabPanel-root css-13xfq8m-MuiTabPanel-root"
      hidden=""
      id="mui-p-12978-P-SENT"
      role="tabpanel"
      style="padding: 1rem 0px;"
    />
  </div>
</div>
`;
